### ISSUE TITLE
Error upgrading from version 3.11.14 (2021082606) to version 3.11.15 (2021082607)

### ISSUE THREAD

**Issue by gerlos** [danmarsden/moodle-mod_attendance#621]
Trying to upgrade attendance plugin from version 3.11.14 (2021082606) to version 3.11.15 (2021082607) I got this error:

```
-->mod_attendance                                                                 
Default exception handler: Si è verificato un errore durante la scrittura nel database Debug: You can't specify target table 'mdl_attendance_log' for update in FROM clause
DELETE FROM mdl_attendance_log                                                    
                 WHERE ID NOT IN (SELECT max(id)                                  
                                    FROM mdl_attendance_log                                                                                                                                                        
                                GROUP BY sessionid, studentid, statusid)                                 
[array (                                                                          
)]                                                                                
Error code: dmlwriteexception                                                     
* line 489 of /lib/dml/moodle_database.php: dml_write_exception thrown            
* line 1167 of /lib/dml/mysqli_native_moodle_database.php: call to moodle_database->query_end()                                                                                                                    
* line 709 of /mod/attendance/db/upgrade.php: call to mysqli_native_moodle_database->execute()                                                                                                                     
* line 877 of /lib/upgradelib.php: call to xmldb_attendance_upgrade()           
* line 576 of /lib/upgradelib.php: call to upgrade_plugins_modules()                                        
* line 1929 of /lib/upgradelib.php: call to upgrade_plugins()                                               
* line 202 of /admin/cli/upgrade.php: call to upgrade_noncore()      
                                                                                  
!!! Si è verificato un errore durante la scrittura nel database !!!                                         
!! You can't specify target table 'mdl_attendance_log' for update in FROM clause                            
DELETE FROM mdl_attendance_log           
                 WHERE ID NOT IN (SELECT max(id)                                  
                                    FROM mdl_attendance_log                                                 
                                GROUP BY sessionid, studentid, statusid)          
[array (                                                                                                    
)]                                                                                
Error code: dmlwriteexception !!                                                                            
!! Stack trace: * line 489 of /lib/dml/moodle_database.php: dml_write_exception thrown                                                                                                                             
* line 1167 of /lib/dml/mysqli_native_moodle_database.php: call to moodle_database-&gt;query_end()                                                                                                                 
* line 709 of /mod/attendance/db/upgrade.php: call to mysqli_native_moodle_database-&gt;execute()                                                                                                                  
* line 877 of /lib/upgradelib.php: call to xmldb_attendance_upgrade()                                       
* line 576 of /lib/upgradelib.php: call to upgrade_plugins_modules()                                        
* line 1929 of /lib/upgradelib.php: call to upgrade_plugins()                                               
* line 202 of /admin/cli/upgrade.php: call to upgrade_noncore()      
 !!                                                                               
```

The issue happens because of this statement in /mod/attendance/db/upgrade.php: 

```
    if ($oldversion < 2021082607) {
        $sql = 'DELETE FROM {attendance_log}
                 WHERE ID NOT IN (SELECT max(id)
                                    FROM {attendance_log}
                                GROUP BY sessionid, studentid, statusid)';
        $DB->execute($sql);

        // Attendance savepoint reached.
        upgrade_mod_savepoint(true, 2021082607, 'attendance');
    }
```

As I read looking around, this happens because MySQL does not allow to DELETE a table's data if you're simultaneously reading that same data with a subquery.

**Comment by dinism** [comment]
Similar error when upgrading 4.0.7 to 4.0.8

**Comment by danmarsden** [comment]
argh - sorry... I've just pushed a quick fix, but may need to come back to this and write a cleanup process for mysql dbs and the duplicate sessions at some point.

**Comment by dinism** [comment]
Solved with version 3.11.16. Thanks a LOT.

**Comment by gerlos** [comment]
This solved the problem for me too, thanks!
